lib/srv/alpnproxy/middleware.go

```diff
@@ -0,0 +1,145 @@
+/*
+Copyright 2022 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package alpnproxy
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"errors"
+	"net"
+
+	"github.com/gravitational/trace"
+	"github.com/jonboulle/clockwork"
+	log "github.com/sirupsen/logrus"
+
+	"github.com/gravitational/teleport/api/client/proto"
+	"github.com/gravitational/teleport/api/utils/keys"
+	"github.com/gravitational/teleport/lib/client"
+	"github.com/gravitational/teleport/lib/tlsca"
+	"github.com/gravitational/teleport/lib/utils"
+)
+
+// LocalProxyMiddleware provides callback functions for LocalProxy.
+type LocalProxyMiddleware interface {
+	// OnNewConnection is a callback triggered when a new downstream connection is
+	// accepted by the local proxy.
+	OnNewConnection(ctx context.Context, lp *LocalProxy, conn net.Conn) error
+	// OnStart is a callback triggered when the local proxy starts.
+	OnStart(ctx context.Context, lp *LocalProxy) error
+}
+
+// DBCertChecker is a middleware that ensures that the local proxy has valid TLS database certs.
+type DBCertChecker struct {
```

Comment:
Personally, I feel the `DBCertChecker` is client specific. It will be great if we can keep this in `tsh` or `lib/client` to avoid `alpnproxy` depending on `lib/client`. What do you guys think?

Reply:
+1

Reply:
moved to lib/client